core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -246,6 +251,38 @@ class BlockManagerMasterEndpoint(
     blockManagerIdByExecutor.get(execId).foreach(removeBlockManager)
   }
 
+  private def recoverLatestRDDBlock(
+      execId: String,
+      excludeExecutors: Seq[String],
+      context: RpcCallContext): Unit = {
+    logDebug(s"Replicating first cached block on $execId")
+    val excluded = excludeExecutors.flatMap(blockManagerIdByExecutor.get)
+    val response: Option[Future[Boolean]] = for {
+      blockManagerId <- blockManagerIdByExecutor.get(execId)
+      info <- blockManagerInfo.get(blockManagerId)
+      blocks = info.cachedBlocks.collect { case r: RDDBlockId => r }
+      // we assume blocks from the latest rdd are most relevant
+      firstBlock <- if (blocks.isEmpty) None else Some(blocks.max[RDDBlockId](Ordering.by(_.rddId)))
+      replicaSet <- blockLocations.asScala.get(firstBlock)
+      // Add 2 below because you need the number of replicas, plus one for the original, plus one
+      // for the new replica.
+      maxReps = replicaSet.size + 2
```

Comment:
doesn't blockLocations already include the original?

Reply:
Oh, you're right it does already include the original. TBH blockManager.replicate subtracts one from the number i put here and I'm not sure what the logic is behind that. I have updated the comment to reflect that we have to add 2 to get a replica on one new executor.